### ISSUE TITLE
Sweep with multiple GPUs at once

### ISSUE THREAD

**Issue by greeneggsandyaml** [facebookresearch/hydra#1331]
Hello authors, thank you for this amazing library!

I recently started to use sweeps and I had a question about running a sweep with multiple runs in parallel. I am not sure where the appropriate place to ask is, so I am asking here. 

I have a script that runs on a single GPU and access to a machine with multiple GPUs (in my case, 4). I would like to run a single sweep in which 4 runs occur in parallel (and then 4 more runs occur after that, and so on). 

I am currently using the optuna plugin, but I would switch to Ax or another framework if that makes this easier. I'm on the latest version of hydra. 

Is this possible with sweeps in hydra? 

Thank you!

**Comment by shagunsodhani** [comment]
Hello @greeneggsandyaml  :) If I understand correctly, you want to run a sweep with atmost 4 jobs at a time? If yes, Ax plugin provides a config key called `max_batch_size` (check the config at the bottom of the page [here](https://hydra.cc/docs/next/plugins/ax_sweeper/)). You can set that to 4.

**Comment by greeneggsandyaml** [comment]
Hello, thank you for the quick response! 

Do you know the best way of making only a single GPU available to each run in the batch? If I simply set `max_batch_size` to 4, then all 4 runs see and use all 4 GPUs, causing overlap. Ideally, each run would only have access to 1 GPU (for example with an environment variable like `CUDA_VISIBLE_DEVICES`).

Thank you again!

**Comment by omry** [comment]
Hydra does not support GPUs directly but it does support [setting environment variables](https://hydra.cc/docs/configure_hydra/job#hydrajobenv_set).
I took a quick look and I don't believe this is flexible enough right now to do what you want though.

I suggest that you try to select the GPU explicitly based on the job number at runtime.
config.yaml:
```yaml
num_gpus: 4
```

my_app.py:
```python
from omegaconf import DictConfig, OmegaConf

import hydra
from hydra.core.hydra_config import HydraConfig

@hydra.main(config_name="config")
def my_app(cfg: DictConfig) -> None:

    print("gpu id:", HydraConfig.get().job.num % cfg.num_gpus)

if __name__ == "__main__":
    my_app()
```
```
(hydra39) omry@2_config_file$ python my_app.py -m '+x=range(1,10)' hydra/hydra_logging=disabled
gpu id: 0
gpu id: 1
gpu id: 2
gpu id: 3
gpu id: 0
gpu id: 1
gpu id: 2
gpu id: 3
gpu id: 0
```

**Comment by jzazo** [comment]
Is there any parameter to start multiple runs with Optuna sweeper rather than Ax by setting a manual param?

@greeneggsandyaml FYI, I haven't tried it, but [pytorch lightning](https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#auto-select-gpus) can select available GPUs automatically if used exclusively.

**Comment by greeneggsandyaml** [comment]
@omry thank you so much for the great response! It looks that this method should work well -- I'll try it out and keep this thread updated. 

For @jzazo, yes `n_jobs` should process multiple jobs in parallel with Optuna.

**Comment by omry** [comment]
@jzazo,
There is a new [Optuna Sweeper Plugin](https://hydra.cc/docs/next/plugins/optuna_sweeper). Give it a shot.

Sweepers are responsible for deciding what to launch, not how.
You can use any launcher that supports parallel execution (basically all launchers except the built-in one).
If you just want local parallel execution you can try the Joblib Launcher.

**Comment by Jasperhino** [comment]
@omry But if I am not mistaking the optuna sweeper is only available for `hydra >= 1.1.0` and the joblib launcher is currently pinned to `hydra = 1.0.*`. How would I go about launching parallel optuna sweeper runs?

Thanks for your work on this amazing library ❤

**Comment by jieru-hu** [comment]
@Jasperhino you can install the ` 1.1.5.dev2` joblib launcher which is compatible with Hydra 1.1

```
pip install hydra-joblib-launcher --pre --upgrade
```

**Comment by mariomeissner** [comment]
@Jasperhino I tried setting `n_jobs=2` but it still didn't execute in parallel. Did you manage to make it work?

**Comment by Jasperhino** [comment]
> @Jasperhino I tried setting `n_jobs=2` but it still didn't execute in parallel. Did you manage to make it work?

I think it worked after installing the `--pre` version as suggested by @jieru-hu

**Comment by Jasha10** [comment]
There is follow-up discussion in https://github.com/facebookresearch/hydra/issues/1974.

**Comment by omry** [comment]
@ds2268
As far as I know, joblib does not know anything about GPUs.
You could use a resource manager like SLURM that can be made aware of the GPUs on each machine, along with the [Submitit Launcher plugin](https://hydra.cc/docs/plugins/submitit_launcher/) which can be launch jobs to SLURM.

**Comment by ds2268** [comment]
I found out that the problem was about memory leakage, like the issue [here](https://github.com/huggingface/transformers/issues/1742).  I have tried clearing the models/schedulers/optimizers (del + gc.collect) and then PyTorch empty_cache, but still couldn't empty the GPU before the next iteration in full. The solution was to use numba like this, to force clean the GPU memory.

```
device = cuda.get_current_device()
device.reset()
cuda.close()
```

It's not a perfect solution but might help others with a similar problem.
